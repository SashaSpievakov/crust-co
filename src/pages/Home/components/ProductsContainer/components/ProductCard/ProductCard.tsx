import { FC, useEffect, useState } from 'react';

import {
  ItemsCountHandler,
  ProductOptions,
} from '../../../../../../components';
import { typeNames } from '../../../../../Pizza/FullProduct/FullProduct';
import {
  Block,
  Bottom,
  CustomLink,
  Image,
  Price,
  Title,
} from './ProductCard.styled';

interface ProductCardProps {
  id: string;
  name: string;
  price: number;
  sizes: number[];
  types: number[];
}

export const ProductCard: FC<ProductCardProps> = ({
  id,
  name,
  price,
  sizes,
  types,
}) => {
  const [activePrice, setActivePrice] = useState<number>(price);
  const [activeSize, setActiveSize] = useState<number>(sizes[0]);
  const [activeType, setActiveType] = useState<number>(0);

  useEffect(() => {
    setActivePrice(price);
  }, [sizes]);

  return (
    <Block>
      <CustomLink to={`/pizzas/${name}`}>
        <picture>
          <source
            srcSet={`./assets/img/pizzas/pizza${id}.webp`}
            type="image/webp"
          />
          <source
            srcSet={`./assets/img/pizzas/pizza${id}.png`}
            type="image/png"
          />
          <Image
            src={`./assets/img/pizzas/pizza${id}.png`}
            alt={`${name} pizza`}
            width="250"
            height="250"
          />
        </picture>
        <Title>{name}</Title>
      </CustomLink>
      <ProductOptions
        price={activePrice}
        sizes={sizes}
        types={types}
        activeSize={activeSize}
        activeType={activeType}
        setActivePrice={setActivePrice}
        setActiveSize={setActiveSize}
        setActiveType={setActiveType}
        typeNames={typeNames}
      />
      <Bottom>
        <Price>{activePrice}$</Price>
        <ItemsCountHandler
          id={id}
          name={name}
          price={activePrice}
          activeSize={activeSize}
          activeType={activeType}
          typeNames={typeNames}
        />
      </Bottom>
    </Block>
  );
};
