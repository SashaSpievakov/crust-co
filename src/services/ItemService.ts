import { createApi, fetchBaseQuery } from '@reduxjs/toolkit/dist/query/react';

import { IPizzaItem } from '../models/IPizzaItem';

const itemAPI = createApi({
  reducerPath: 'itemAPI',
  baseQuery: fetchBaseQuery({
    baseUrl: `${process.env.REACT_APP_API_URL}/api`,
  }),
  tagTypes: ['Item'],
  endpoints: (build) => ({
    fetchItem: build.query<IPizzaItem, string>({
      query: (id) => ({
        url: `/pizzas/${id}`,
      }),
      providesTags: () => ['Item'],
    }),
  }),
});

export default itemAPI;
