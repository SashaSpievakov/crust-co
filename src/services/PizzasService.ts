import { createApi, fetchBaseQuery } from '@reduxjs/toolkit/dist/query/react';

import { IPizzaItem } from '../models/IPizzaItem';

interface IFetchItems {
  activeCategory: number;
  sortedPropertyName: string;
}

const pizzasAPI = createApi({
  reducerPath: 'pizzasAPI',
  baseQuery: fetchBaseQuery({
    baseUrl: 'https://6344adb1dcae733e8fe3067a.mockapi.io/',
  }),
  tagTypes: ['Pizzas'],
  endpoints: (build) => ({
    fetchPizzas: build.query<IPizzaItem[], IFetchItems>({
      query: ({ activeCategory, sortedPropertyName }) => ({
        url: `pizza-items?${
          activeCategory > 0 ? `category=${activeCategory}&` : ''
        }sortBy=${sortedPropertyName}`,
      }),
      providesTags: () => ['Pizzas'],
    }),
  }),
});

export default pizzasAPI;
