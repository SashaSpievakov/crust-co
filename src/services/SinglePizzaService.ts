import { createApi, fetchBaseQuery } from '@reduxjs/toolkit/dist/query/react';

import { IPizzaItem } from '../models/IPizzaItem';

export const pizzaAPI = createApi({
  reducerPath: 'pizzaPI',
  baseQuery: fetchBaseQuery({
    baseUrl: `${process.env.REACT_APP_API_URL}/api`,
  }),
  tagTypes: ['Pizza'],
  endpoints: (build) => ({
    fetchItem: build.query<IPizzaItem, string>({
      query: (name) => ({
        url: `/pizzas/${name}`,
      }),
      providesTags: (result) =>
        result ? [{ type: 'Pizza', id: result.name }] : ['Pizza'],
    }),
  }),
});
