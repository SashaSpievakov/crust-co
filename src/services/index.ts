export { pizzasAPI } from './PizzasService';
export { pizzaAPI } from './SinglePizzaService';
