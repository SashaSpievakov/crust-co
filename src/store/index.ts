export * from './slices';
export type { AppDispatch, AppStore, RootState } from './store';
export { setupStore } from './store';
