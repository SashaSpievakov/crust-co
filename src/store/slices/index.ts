export type { CartSliceState } from './cart';
export { addItem, clearCart, removeItem, removeItems } from './cart';
export { selectIsLight, setTheme } from './theme';
