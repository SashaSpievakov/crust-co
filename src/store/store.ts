import {
  combineReducers,
  configureStore,
  PreloadedState,
} from '@reduxjs/toolkit';

import categorySlice from './slices/category/reducer/categoryReducer';
import sortSlice from './slices/sort/reducer/sortReducer';
import searchSlice from './slices/search/reducer/searchValueReducer';
import theme from './slices/theme/reducer/themeReducer';
import cart from './slices/cart/reducer/cartReducer';
import itemAPI from '../services/ItemService';
import pizzasAPI from '../services/PizzasService';

const rootReducer = combineReducers({
  activeCategory: categorySlice,
  activeSort: sortSlice,
  searchValue: searchSlice,
  theme,
  cart,
  [itemAPI.reducerPath]: itemAPI.reducer,
  [pizzasAPI.reducerPath]: pizzasAPI.reducer,
});

export const setupStore = (preloadedState?: PreloadedState<RootState>) => {
  return configureStore({
    reducer: rootReducer,
    middleware: (getDefaultMiddleware) =>
      getDefaultMiddleware()
        .concat(itemAPI.middleware)
        .concat(pizzasAPI.middleware),
    preloadedState,
  });
};

export type RootState = ReturnType<typeof rootReducer>;
export type AppStore = ReturnType<typeof setupStore>;
export type AppDispatch = AppStore['dispatch'];
