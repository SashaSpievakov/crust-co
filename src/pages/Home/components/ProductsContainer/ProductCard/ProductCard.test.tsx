import '@testing-library/jest-dom';

import { screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { pizzaAPI } from '../../../../../services';
import { setupStore } from '../../../../../store/store';
import {
  rendererWithProviders,
  renderWithProvidersAndRoutes,
} from '../../../../../tests/helpers';
import { mockItem, testServer } from '../../../../../tests/mocks';
import { ProductCard } from './ProductCard';

describe('ProductCard Tests', () => {
  test('renders the ProductCard component', () => {
    const snapshot = rendererWithProviders(<ProductCard {...mockItem} />);
    expect(snapshot).toMatchSnapshot();
  });

  describe('checks the price changing', () => {
    beforeEach(() => {
      renderWithProvidersAndRoutes(<ProductCard {...mockItem} />);
    });

    test('checks type click', () => {
      const typeItem = screen.getByText(/thin/i);
      const price = screen.getByText(/14\$/i);

      userEvent.click(typeItem);

      expect(price).toHaveTextContent('15$');
    });

    test('checks size click', () => {
      const sizeItem = screen.getByText(/16 inch/i);
      const price = screen.getByText(/14\$/i);

      userEvent.click(sizeItem);

      expect(price).toHaveTextContent('18$');
    });

    test('checks type and size click', () => {
      const typeItem = screen.getByText(/thin/i);
      const sizeItem = screen.getByText(/14 inch/i);
      const price = screen.getByText(/14\$/i);

      userEvent.click(typeItem);
      userEvent.click(sizeItem);

      expect(price).toHaveTextContent('17$');
    });
  });

  describe('checks rtk query hook', () => {
    beforeAll(() => {
      testServer.listen();
    });

    afterAll(() => {
      testServer.close();
    });

    afterEach(() => {
      testServer.resetHandlers();
      setupStore().dispatch(pizzaAPI.util.resetApiState());
    });

    test('checks link redirect to the Item page', async () => {
      renderWithProvidersAndRoutes(
        <ProductCard {...mockItem} />,
        true,
        '/cart',
      );
      const link = screen.getByRole('heading', {
        name: /chicken curry/i,
      });

      userEvent.click(link);

      expect(
        await screen.findByText(
          /Lorem, ipsum dolor sit amet consectetur adipisicing elit. Aliquid nisi aspernatur debitis quod/i,
        ),
      ).toBeInTheDocument();
    });
  });
});
