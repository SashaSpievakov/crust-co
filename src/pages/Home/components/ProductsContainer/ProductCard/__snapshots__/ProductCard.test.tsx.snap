// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductCard Tests renders the ProductCard component 1`] = `
Object {
  "children": Array [
    <a
      className="sc-fvEvSO jSMwtJ"
      href="/pizzas/Chicken Curry"
      onClick={[Function]}
    >
      <picture>
        <source
          srcSet="./assets/img/pizzas/pizza9.webp"
          type="image/webp"
        />
        <source
          srcSet="./assets/img/pizzas/pizza9.png"
          type="image/png"
        />
        <img
          alt="Chicken Curry pizza"
          className="sc-fcCDlF figEwF"
          height="250"
          src="./assets/img/pizzas/pizza9.png"
          width="250"
        />
      </picture>
      <h3
        className="sc-GKYbw gOnPJi"
      >
        Chicken Curry
      </h3>
    </a>,
    <div
      className="sc-eKJbhj gmULXV"
    >
      <ul>
        <li
          aria-current={true}
          className="sc-bTTELM hWocCQ"
          onClick={[Function]}
          onKeyDown={[Function]}
          tabIndex={0}
        >
          traditional
        </li>
        <li
          aria-current={false}
          className="sc-bTTELM fNZKrT"
          onClick={[Function]}
          onKeyDown={[Function]}
          tabIndex={0}
        >
          thin
        </li>
      </ul>
      <ul>
        <li
          aria-current={true}
          className="sc-bTTELM hWocCQ"
          onClick={[Function]}
          onKeyDown={[Function]}
          tabIndex={0}
        >
          12
           inch
        </li>
        <li
          aria-current={false}
          className="sc-bTTELM fNZKrT"
          onClick={[Function]}
          onKeyDown={[Function]}
          tabIndex={0}
        >
          14
           inch
        </li>
        <li
          aria-current={false}
          className="sc-bTTELM fNZKrT"
          onClick={[Function]}
          onKeyDown={[Function]}
          tabIndex={0}
        >
          16
           inch
        </li>
      </ul>
    </div>,
    <section
      className="sc-gsGlKL dFNTzs"
    >
      <h3
        className="sc-bWOGAC jKsOjm"
      >
        14
        $
      </h3>
      <button
        className="sc-bcXHqe sc-eDvSVe sc-gKPRtg glVyWT fwWQTF gmBSoV"
        onClick={[Function]}
      >
        <svg
          className="sc-hlLBRy dBNJHM"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 16 16"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M8 0a1 1 0 0 1 1 1v6h6a1 1 0 1 1 0 2H9v6a1 1 0 1 1-2 0V9H1a1 1 0 0 1 0-2h6V1a1 1 0 0 1 1-1z"
          />
        </svg>
        <span>
          Add
        </span>
      </button>
    </section>,
  ],
  "props": Object {
    "className": "sc-jRwbcX giLUud",
  },
  "type": "article",
}
`;
